File: example/javalib/script/2-crawler/Crawler.java
//| mvnDeps:
//| - info.picocli:picocli:4.7.6
//| - com.squareup.okhttp3:okhttp:4.12.0
//| - com.fasterxml.jackson.core:jackson-databind:2.17.2

import com.fasterxml.jackson.databind.*;
import com.fasterxml.jackson.core.util.DefaultIndenter;
import com.fasterxml.jackson.core.util.DefaultPrettyPrinter;
import okhttp3.*;
import picocli.CommandLine;
import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.concurrent.Callable;

@CommandLine.Command(name = "Crawler", mixinStandardHelpOptions = true)
public class Crawler implements Callable<Integer> {

  @CommandLine.Option(
    names = {"--start-article"},
    required = true,
    description = "Starting article title"
  )
  private String startArticle;

  @CommandLine.Option(
    names = {"--depth"},
    required = true,
    description = "Depth of crawl"
  )
  private int depth;

  private static final OkHttpClient client = new OkHttpClient();
  private static final ObjectMapper mapper = new ObjectMapper();

  public static List<String> fetchLinks(String title) throws IOException {
    var url = new HttpUrl.Builder()
      .scheme("https")
      .host("en.wikipedia.org")
      .addPathSegments("w/api.php")
      .addQueryParameter("action", "query")
      .addQueryParameter("titles", title)
      .addQueryParameter("prop", "links")
      .addQueryParameter("format", "json")
      .build();

    var request = new Request.Builder()
      .url(url)
      .header("User-Agent", "WikiFetcherBot/1.0 (https://example.com; contact@example.com)")
      .build();

    try (var response = client.newCall(request).execute()) {
      if (!response.isSuccessful())
        throw new IOException("Unexpected code " + response);

      var root = mapper.readTree(response.body().byteStream());
      var pages = root.path("query").path("pages");
      var links = new ArrayList<String>();

      for (var it = pages.elements(); it.hasNext();) {
        var linkArr = it.next().get("links");
        if (linkArr != null && linkArr.isArray()) {
          for (var link : linkArr) {
            var titleNode = link.get("title");
            if (titleNode != null) links.add(titleNode.asText());
          }
        }
      }
      return links;
    }
  }

  @Override
  public Integer call() throws Exception {
    var seen = new HashSet<String>();
    var current = new HashSet<String>();
    seen.add(startArticle);
    current.add(startArticle);

    for (int i = 0; i < depth; i++) {
      var next = new HashSet<String>();
      for (var article : current) {
        for (var link : fetchLinks(article)) {
          if (!seen.contains(link)) next.add(link);
        }
      }
      seen.addAll(next);
      current = next;
    }

    var output = Paths.get("fetched.json");
    try (var w = Files.newBufferedWriter(output)) {
      var printer = new DefaultPrettyPrinter();
      printer.indentArraysWith(new DefaultIndenter("    ", "\n"));
      printer.indentObjectsWith(new DefaultIndenter("    ", "\n"));
      mapper.writer(printer).writeValue(w, seen);
    }
    return 0;
  }

  public static void main(String[] args) {
    System.exit(new CommandLine(new Crawler()).execute(args));
  }
}

File: example/kotlinlib/script/2-crawler/Crawler.kt
//| mvnDeps:
//| - com.github.ajalt.clikt:clikt:5.0.3
//| - com.konghq:unirest-java:3.14.5
//| - org.jetbrains.kotlinx:kotlinx-serialization-json:1.7.3

import com.github.ajalt.clikt.core.CliktCommand
import com.github.ajalt.clikt.core.main
import com.github.ajalt.clikt.parameters.options.*
import com.github.ajalt.clikt.parameters.types.int
import kotlinx.serialization.json.*
import kong.unirest.Unirest
import java.nio.file.*

fun fetchLinks(title: String): List<String> {
    val response = Unirest.get("https://en.wikipedia.org/w/api.php")
        .queryString("action", "query")
        .queryString("titles", title)
        .queryString("prop", "links")
        .queryString("format", "json")
        .header("User-Agent", "WikiFetcherBot/1.0 (https://example.com; contact@example.com)")
        .asString()

    if (!response.isSuccess) return emptyList()

    val json = Json.parseToJsonElement(response.body).jsonObject
    val pages = json["query"]?.jsonObject?.get("pages")?.jsonObject ?: return emptyList()
    return pages.values.flatMap { page ->
        page.jsonObject["links"]
            ?.jsonArray
            ?.mapNotNull { it.jsonObject["title"]?.jsonPrimitive?.content }
            ?: emptyList()
    }
}

class Crawler : CliktCommand(name = "wiki-fetcher") {
    val startArticle by option(help = "Starting Wikipedia article").required()
    val depth by option(help = "Depth of link traversal").int().required()

    override fun run() {
        var seen = mutableSetOf(startArticle)
        var current = mutableSetOf(startArticle)

        repeat(depth) {
            val next = current.flatMap { fetchLinks(it) }.toSet()
            current = (next - seen).toMutableSet()
            seen += current
        }

        val jsonOut = Json { prettyPrint = true }
            .encodeToString(JsonElement.serializer(), JsonArray(seen.map { JsonPrimitive(it) }))
        Files.writeString(Paths.get("fetched.json"), jsonOut)
    }
}

fun main(args: Array<String>) = Crawler().main(args)